tests/recipes/test_full_finetune_single_device.py

```diff
@@ -0,0 +1,235 @@
+        ckpt_path = Path(CKPT_MODEL_PATHS[ckpt])
+        ckpt_dir = ckpt_path.parent
+        log_file = gen_log_file_name(tmpdir)
+
+        # Config file needed for model conversion.
+        # Create a second copy for training resume
+        write_hf_ckpt_config(ckpt_dir)
+        write_hf_ckpt_config(tmpdir)
+
+        # Train for two epochs
+        cmd_1 = f"""
+        tune full_finetune_single_device
+            --config full_finetune_single_device \
+            output_dir={tmpdir} \
+            checkpointer._component_=torchtune.utils.FullModelHFCheckpointer \
+            checkpointer.checkpoint_dir='{ckpt_dir}' \
+            checkpointer.checkpoint_files=[{ckpt_path}]\
+            checkpointer.output_dir={tmpdir} \
+            checkpointer.model_type=LLAMA2 \
+        """.split()
+
+        model_config = llama2_test_config()
+        cmd_1 = cmd_1 + self._get_test_config_overrides() + model_config
+
+        monkeypatch.setattr(sys, "argv", cmd_1)
+        with pytest.raises(SystemExit):
+            runpy.run_path(TUNE_PATH, run_name="__main__")
+
+        # Resume training
+        cmd_2 = f"""
+        tune full_finetune_single_device
+            --config full_finetune_single_device \
+            output_dir={tmpdir} \
+            checkpointer._component_=torchtune.utils.FullModelHFCheckpointer \
+            checkpointer.checkpoint_dir={tmpdir} \
+            checkpointer.checkpoint_files=[{os.path.join(tmpdir, "hf_model_0001_0.pt")}]\
+            checkpointer.recipe_checkpoint={os.path.join(tmpdir, "recipe_state.pt")}
+            checkpointer.output_dir={tmpdir} \
+            checkpointer.model_type=LLAMA2 \
+            resume_from_checkpoint=True \
+            metric_logger.filename={log_file} \
+        """.split()
+
+        cmd_2 = cmd_2 + self._get_test_config_overrides() + model_config
+
+        monkeypatch.setattr(sys, "argv", cmd_2)
+        with pytest.raises(SystemExit):
+            runpy.run_path(TUNE_PATH, run_name="__main__")
+
+        expected_loss_values = self._fetch_expected_loss_values()[2:]
+
+        loss_values = get_loss_values_from_metric_logger(log_file)
+        torch.testing.assert_close(
+            loss_values, expected_loss_values, rtol=1e-4, atol=1e-4
+        )
+
+
+class TestFullFinetuneSingleDeviceGradientAccumulation:
+    def _get_test_config_overrides(self):
+        return [
+            "device=cpu",
+            "dtype=fp32",
+            "enable_activation_checkpointing=False",
+            "tokenizer.path=/tmp/test-artifacts/tokenizer.model",
+            "dataset=tests.recipes.utils.DummyDataset",
+            "dataset.train_on_input=False",
+            "seed=9",
+            "epochs=1",
+            "max_steps_per_epoch=1",
+            "optimizer=torch.optim.AdamW",
+            "optimizer.lr=2e-5",
+            "log_every_n_steps=1",
+        ]
+
+    @pytest.mark.integration_test
+    @pytest.mark.parametrize("full_batch_size, micro_batch_size", [(2, 1), (4, 1)])
```

Comment:
What's the runtime of this test? Is there value on testing on multiple pairs of full, micro batch size?

Reply:
Just ran this one and it actually only took about 5 seconds (probably because the model is small and we're on single device). Re multiple pairs it is kind of a holdover from the original version, and actually there was a point where the `(2, 1)` case was passing as a false positive (negative?) due to too high of a tolerance. But now that we've cleaned it up maybe we can remove it? Honestly given that the test is relatively quick I don't have a strong preference here either way.

Reply:
If we feel confident, I'd be up for chipping away at the total running time by removing some tests

Reply:
Sounds good, will remove the (2, 1) case then.